### PR TITLE
EVG-2862 Implement Notifications Subscription and Subscriber model

### PR DESCRIPTION
**Pull request by richardsamuels**

### CODE REVIEW

File: model/event/subscribers.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	}
+
+	// TODO registry?
+	switch s.Type {
+	case "github_pull_request":
+		s.Target = githubPullRequestSubscriber{}
+	case "webhook":
+		s.Target = webhookSubscriber{}
+	default:
+		s.Target = string(temp.Target)
```
Comment (john-m-liu) [inline model/event/subscribers.go]: Does it make sense to have this case be here? Do we do something sane if this is a string?

File: model/event/subscriptions.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+			},
+		},
+		{
+			"$project": bson.M{
+				subscriptionSubscriberKey:     1,
+				subscriptionRegexSelectorsKey: 1,
+				"keep": bson.M{
+					"$and": []bson.M{
+						{
+							"$ne": []interface{}{
```
Comment (john-m-liu) [inline model/event/subscriptions.go]: should be a []bson.M

File: model/event/subscriptions.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+								"$" + subscriptionSelectorsKey,
+								[]interface{}{},
+							},
+						},
+						{
+							"$setIsSubset": []interface{}{"$" + subscriptionSelectorsKey, selectors},
```
Comment (john-m-liu) [inline model/event/subscriptions.go]: should be a []bson.M

File: model/event/subscribers.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	if err := raw.Unmarshal(&temp); err != nil {
+		return errors.Wrap(err, "can't unmarshal subscriber data")
+	}
+	s.Type = temp.Type
+
+	if len(s.Type) == 0 {
+		return errors.New("could not find subscriber type")
+	}
+
+	// TODO registry?
```
Comment (john-m-liu) [inline model/event/subscribers.go]: Might want to make this a registry now if we're planning on adding new subscriber types soon

File: model/event/subscribers.go (hunk 1 above)
Comment (richardsamuels) [reply model/event/subscribers.go]: Yes, the default case is that the subscriber info can be rendered in just a string, such as emails, slack users, jira projects

File: model/event/subscriptions.go (hunk 2 above)
Comment (richardsamuels) [reply model/event/subscriptions.go]: this is a slice containing a string, and an empty slice of something. This has to be []inteface{}.

File: model/event/subscriptions.go (hunk 3 above)
Comment (richardsamuels) [reply model/event/subscriptions.go]: this is a slice containing a string, and a slice of strings. This has to be []inteface{}.

File: model/event/subscribers.go (hunk 4 above)
Comment (richardsamuels) [reply model/event/subscribers.go]: We're not, but i'm not opposed to making this a registry

File: model/event/subscribers.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import (
+	"net/url"
+
+	"github.com/mongodb/anser/bsonutil"
+	"github.com/pkg/errors"
+	"gopkg.in/mgo.v2/bson"
+)
+
+const (
+	ResourceTypeNotification = "notification"
```
Comment (bsamek) [inline model/event/subscribers.go]: nit: can this be private?

File: model/event/subscribers.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	temp := struct {
+		Type   string `bson:"type"`
+		Target []byte `bson:"target"`
+	}{}
+	if err := raw.Unmarshal(&temp); err != nil {
+		return errors.Wrap(err, "can't unmarshal subscriber data")
+	}
+	s.Type = temp.Type
+
+	if len(s.Type) == 0 {
```
Comment (bsamek) [inline model/event/subscribers.go]: nit: check length of temp.Type and return err before assigning to s.Type

File: model/event/subscribers.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+		return errors.New("could not find subscriber type")
+	}
+
+	// TODO registry?
+	switch s.Type {
+	case "github_pull_request":
```
Comment (bsamek) [inline model/event/subscribers.go]: constants here and "webhook"

File: scripts/indexes.js [hunk 8]
```diff
@@ -115,3 +115,6 @@ db.patch_intents.ensureIndex({ "msg_id" : 1 }, { "unique" : true, "sparse": true
 
 //======github_hooks======//
 db.github_hooks.ensureIndex({ "owner" : 1, "repo" : 1 }, { "unique": true })
+
+//======subscriptions======//
+db.subscriptions.ensureIndex({ "type": 1, "trigger" : 1, "selectors": 1})
```
Comment (bsamek) [inline scripts/indexes.js]: make an MCI ticket in the epic for this

File: model/event/subscribers.go (hunk 6 above)
Comment (richardsamuels) [reply model/event/subscribers.go]: done

File: scripts/indexes.js (hunk 8 above)
Comment (richardsamuels) [reply scripts/indexes.js]: done

File: model/event/subscribers.go (hunk 7 above)
Comment (richardsamuels) [reply model/event/subscribers.go]: done

File: model/event/subscribers.go (hunk 5 above)
Comment (richardsamuels) [reply model/event/subscribers.go]: That shouldn't event be here, removed it.

**Pull request merged**